app/services/push_notification/http_push.rb

```diff
@@ -0,0 +1,72 @@
+module PushNotification
+  # Delivers SETs (Security Event Tokens) via the HTTP Push protocol
+  # https://tools.ietf.org/html/draft-ietf-secevent-http-push-00
+  class HttpPush
+    include Rails.application.routes.url_helpers
+
+    attr_reader :event
+
+    # shorthand for creating an instance then calling #deliver, easier to stub
+    def self.deliver(event, now: Time.zone.now)
+      new(event, now: now).deliver
+    end
+
+    def initialize(event, now: Time.zone.now)
+      @event = event
+      @now = now
+    end
+
+    def deliver
+      event.user.
```

Comment:
May I suggest to move  this knowledge about `user` & `Identity` to `PushNotification::IdentifierRecycledEvent` ?
So `#deliver` would know nothing about what makes suitable to deliver?

```ruby
def deliver
  event.push_notifications_urls.each do |service_provider|
     deliver_one(service_provider)
  end
end
```

I understand `IdentifierRecycledEvent` is more like a Value object, but it would be interesting that `HttpPush` be a pure http-data class.

Reply:
Thanks! Definitely something I considered.

Like you noticed, trying to keep the event classes more like value objects. If we find ourselves doing varied kinds of delivery later, maybe we can refactor out the HTTP part into something else.

One other reason I feel like this logic should be here is that we plan to have other event types in the near future, and having the delivery logic here makes it consistent. If we had this logic in event objects, we'd have to worry about different event objects behaving similarly, so we'd have to create subclasses (which can be hard to find, and could still override and change things).

For the time being I think I'm going to keep this structure

Reply:
Thanks for the interesting explanation :)